specs/AppLifecycle/StateNotifications/AppLifecycle StateNotifications.md

```diff
@@ -0,0 +1,389 @@
+
+Classic Win32 apps can use either the WM model and/or the callback model. The callback model is more
+consistent with other app types - including not only modern UWP apps, but also traditional MFC,
+Winforms and WPF apps. Also, the callback model does not prescribe a windowing or threading model,
+which is better for headless/background components. Given this, since all app types can work with
+the callback model, but only classic Win32 apps use the WM model, the Reunion API will use the
+callback model for Cobalt. Additional WM-based behavior can be added in a later release, if customer
+demand indicates it would provide significant benefit.
+
+The net is that the Reunion PowerManager API will be a combination of WinRT PowerManager and Win32
+PowerSettingRegisterNotification. We will define new WinRT types (in
+Microsoft.ApplicationModel.PowerManager) that provide event handlers for apps to use.
+
+## Examples
+
+### Subscribing and responding to events
+
+The power notifications are relatively simple. The app provides a callback function for each
+notification it wants to receive, and when the event is raised the system invokes that callback. The
+app could start listening for the notifications at any time; it would most often do this early in
+its life, perhaps in WinMain - but the AppLifecycle mechanism does not impose any constraints on
+this, and the app is free to subscribe or unsubscribe at any time.
+
+```c++
+int APIENTRY wWinMain(
+    _In_ HINSTANCE hInstance, _In_opt_ HINSTANCE hPrevInstance,
+    _In_ LPWSTR lpCmdLine, _In_ int nCmdShow)
+{
+    // Optionally, register callbacks for power/battery state changes.
+    PowerManager::BatteryStatusChanged([](auto&&...)
+        { OnBatteryStateChange(); });
+    PowerManager::PowerSupplyStatusChanged([](auto&&...)
+        { OnPowerStateChange(); });
+    PowerManager::EnergySaverStatusChanged([](auto&&...)
+        { OnEnergySaverStateChange(); });
+    PowerManager::RemainingChargePercentChanged([](auto&&...)
+        { OnRemainingChargePercentChange(); });
+    PowerManager::RemainingDischargeTimeChanged([](auto&&...)
+        { OnRemainingDischargeTimeChange(); });
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Standard Win32 window configuration/creation and message pump:
+    // ie, whatever the app would normally do - nothing new here.
+    LoadStringW(hInstance, IDS_APP_TITLE, szTitle, MAX_LOADSTRING);
+    LoadStringW(hInstance, IDC_CLASSIC, szWindowClass, MAX_LOADSTRING);
+    MyRegisterClass(hInstance);
+    if (!InitInstance(hInstance, nCmdShow))
+    {
+        return FALSE;
+    }
+    MSG msg;
+    while (GetMessage(&msg, nullptr, 0, 0))
+    {
+        TranslateMessage(&msg);
+        DispatchMessage(&msg);
+    }
+    return (int)msg.wParam;
+}
+```
+
+In the app-defined callback, the app can do whatever it likes. Typically, it would fetch the current
```

Comment:
Presumably the callback is invoked from an arbitrary thread (properly marshaled, to respect the agility of the delegate), and apps need to apply appropriate synchronization as required. Should document this.

Reply:
RECOMMEND: Document that these are regular WinRT event delegate invocations. See also https://github.com/microsoft/xlang/issues/673

Reply:
Added further description, plus the link.